### ISSUE TITLE
:sleep in async callback risks unpredictable state

### ISSUE THREAD

**Issue by bhcleek** [neovim/neovim#10018]
- `nvim --version`: 0.3.4

### Steps to reproduce using `nvim -u NORC`

Run two plugins. The first should call `sleep` periodically. The second should process output from an async job that produces output similar to `gopls`. i.e. json responses that are preceded with a line that indicates how long the following json message (which may span multiple lines) is in bytes.

ref: https://github.com/fatih/vim-go/issues/2274#issuecomment-490647086

### Actual behaviour
The stdout callback may be called before its previous invocation is finished, resulting in lines effectively being interleaved and making it difficult to process the json messages reliably.

### Expected behaviour
The stdout callback should not be called before its previous invocation is finished so that line-oriented protocols can be interpreted reliably.

**Comment by justinmk** [comment]
see `:help channel-lines` for explanation and boilerplate solution, i.e.:

```vim
          let s:chunks = ['']
          func! s:on_event(job_id, data, event) dict
            let s:chunks[-1] .= a:data[0]
            call extend(s:chunks, a:data[1:])
          endf
```

Alternatively use `std{out,err}_buffered`, see `:help  `on_stdout`

**Comment by bhcleek** [comment]
I've read the docs and am familiar with how Neovim works.

That doesn't solve the problem at all. None of those operations are atomic, and therefore race conditions can exist where the `on_event` is called twice concurrently, possibly results in the second invocation extending `s:chunks` before the first invocation does.

**Comment by justinmk** [comment]
> The stdout callback should not be called before its previous invocation is finished so that line-oriented protocols can be interpreted reliably.

What does this mean? A callback does not have any concept of its "previous invocation".

**Comment by bhcleek** [comment]
> What does this mean? A callback does not have any concept of its "previous invocation".

The callback can be invoked multiple times. It should not be invoked again until a a prior invocation has completed.

**Comment by justinmk** [comment]
> That doesn't solve the problem at all. None of those operations are atomic, and therefore race conditions can exist where the `on_event` is called twice concurrently, 

If by "concurrently"  you mean "in parallel", that's not possible, Nvim is single-threaded and VimL function execution is never interrupted or pre-empted for some other VimL function. 

> The callback can be invoked multiple times. It should not be invoked again until a a prior invocation has completed.

Nvim and Vim are single-threaded, there is no case where the callback would be invoked before some other invocation is incomplete.

**Comment by bhcleek** [comment]
That's what I thought before I read https://github.com/fatih/vim-go/issues/2274#issuecomment-490647086, but I'm at a loss to explain what we're observing if `sleep` doesn't somehow cause Neovim to evaluate a job's output and invoke the callback.

**Comment by justinmk** [comment]
You're right, this makes sense to me:

> * during the execution of `a:callback` for the third element (the one with the location of the GoDef symbol), a new go file is opened, more requests to gopls are made, numerous autocommands are triggered and one such autocommand belongs to a different vim plugin called gitgutter. In that code, it calls [`sleep 5m`](https://github.com/airblade/vim-gitgutter/blob/1d422b9f98194e38bc56e54192c9bc66d95c21f1/autoload/gitgutter/diff.vim#L72).

> 3\. This is where the bug manifests: the sleep call prompts neovim to process more output from gopls, but it does so before the while loop in (5) got a chance to make `call a:callback(l:msg)` calls for the last two elements.

**Comment by bfredl** [comment]
Calling `sleep` inside an async handler is not recommended, it can lead to all sort of issues. But if someone does, we probably should at least block callbacks from the same channel.

**Comment by bhcleek** [comment]
We're not calling sleep from within the handler. It's a completely separate plugin that's calling sleep.

**Comment by justinmk** [comment]
I guess we should add "invoking autocmds" to the list of things that are risky in async handlers :)

**Comment by bhcleek** [comment]
Sure, but even in that case, we're not explicitly invoking autocmds.

**Comment by justinmk** [comment]
Understood.

> Calling `sleep` inside an async handler is not recommended, it can lead to all sort of issues. But if someone does, we probably should at least block callbacks from the same channel.

Why don't we disable sleep entirely in callbacks?

**Comment by bhcleek** [comment]
> Sure, but even in that case, we're not explicitly invoking autocmds.

I need to correct that.

The autocmds in this case are being called because new buffers are being opened in response to the the messages being processed from the async job.

Perhaps the solution here is for gitgutter to use `timer_start` instead of sleep 🤔

**Comment by bhcleek** [comment]
Thanks for talking this through. I realize that there may not be much to be done here other than making it clear in docs that `sleep` can be dangerous.

If there is a solution, though, I'm sure the community would appreciate it it. Perhaps it's as simple as not invoking a callback again due to a `sleep` if a prior invocation of that callback is already in-flight?

**Comment by justinmk** [comment]
PoC in https://github.com/neovim/neovim/pull/10019. Would be useful if someone could check that it doesn't make things worse.

**Comment by bfredl** [comment]
> I guess we should add "invoking autocmds" to the list of things that are risky in async handlers :)

As pointed out, too many useful things will invoke autocmds. Buffer autocmds are not async events. I think we should rather say say using `sleep` in an autocmd should be considered dubious. 

Disabling event processing in `sleep` will just cause unmodifed gitgutter to go in a deadlock instead. The fix would be for gitgutter to use `jobwait()`, or whatever scoped event processing function that is relevant.

**Comment by bfredl** [comment]
Also, `sleep` and `getchar()` are not the only ways deferred events can be invoked. I think the alternative is while `on_stdout` on a specific channel is running, then callbacks on the _same_ channel will be blocked.

We should do the same for timers (which vim does, missing vim-patch?).

**Comment by justinmk** [comment]
> while `on_stdout` on a specific channel is running, then callbacks on the _same_ channel will be blocked.

How would that work? I can think of one approach: similar to `ch_before_blocking_events` https://github.com/neovim/neovim/blob/9420a2127fbd6a7dc28ce4115da47bb56c57e235/src/nvim/msgpack_rpc/channel.c#L49

we could have a private multiqueue per-channel. This is pretty ugly though. 

- Calling `rstream_stop` in `on_channel_event` is too late, there may already be pending channel events which could get processed by `:sleep`.
- Maybe `multiqueue_process_events` could take a predicate and it skips events matching the predicate. That makes the system harder to understand though.

> same for timers (which vim does, missing vim-patch?).

can you point to the vim code?

**Comment by bfredl** [comment]
I think It can be done quite easily without messing with queues or anything, we already have throttling logic. I will do a prototype.

**Comment by bfredl** [comment]
I did some larger refactors at the same time :) https://github.com/neovim/neovim/pull/10021

> can you point to the vim code?

@justinmk I realized this is a consequence of the overall different timer semantics in vim.  https://github.com/neovim/neovim/issues/6829 .

**Comment by justinmk** [comment]
closed by #10021
